### ISSUE TITLE
runc: devices_unix.go: Race condition in getDevices()

### ISSUE THREAD

**Issue by agunnerson-ibm** [moby/moby#32667]
In `devices_unix.go` from runc, `getDevices()` scans for files in `/dev` and then calls `os.Lstat()` on them. A race condition is introduced if anything in `/dev` is deleted between getting the list of files and accessing them.

This has been fixed upstream with https://github.com/opencontainers/runc/commit/8f55948aa5c97472c4a8ef6664b10ed662ce1df5 but it appears that the change has not made it to docker yet.

This can be easily reproduced by running a bunch of `--privileged` docker containers at once:

```sh
for i in {0..100}; do docker run --rm --privileged busybox echo "test_${i}" & done; wait
```

On one of our systems, this will result in the following. Note the lstat failures. The unmount failures are probably an unrelated issue.

```sh
[root@[REDACTED] andrew.gunnerson]# for i in {0..100}; do docker run --rm --privileged busybox echo "test_${i}" & done; wait
[1] 21515
[2] 21516
[3] 21517
[4] 21518
[5] 21519
[6] 21520
[7] 21521
[8] 21522
[9] 21523
[10] 21524
[11] 21525
[12] 21526
[13] 21527
[14] 21529
[15] 21530
[16] 21531
[17] 21532
[18] 21533
[19] 21534
[20] 21535
[21] 21536
[22] 21537
[23] 21538
[24] 21539
[25] 21540
[26] 21541
[27] 21542
[28] 21544
[29] 21545
[30] 21546
[31] 21547
[32] 21548
[33] 21549
[34] 21550
[35] 21551
[36] 21552
[37] 21553
[38] 21554
[39] 21555
[40] 21556
[41] 21557
[42] 21558
[43] 21559
[44] 21563
[45] 21567
[46] 21570
[47] 21576
[48] 21579
[49] 21583
[50] 21586
[51] 21588
[52] 21591
[53] 21593
[54] 21595
[55] 21598
[56] 21601
[57] 21603
[58] 21607
[59] 21611
[60] 21614
[61] 21616
[62] 21618
[63] 21623
[64] 21627
[65] 21628
[66] 21630
[67] 21632
[68] 21641
[69] 21643
[70] 21646
[71] 21647
[72] 21648
[73] 21653
[74] 21657
[75] 21660
[76] 21663
[77] 21664
[78] 21666
[79] 21669
[80] 21673
[81] 21698
[82] 21699
[83] 21701
[84] 21704
[85] 21705
[86] 21708
[87] 21710
[88] 21713
[89] 21715
[90] 21717
[91] 21718
[92] 21719
[93] 21720
[94] 21722
[95] 21725
[96] 21729
[97] 21731
[98] 21732
[99] 21735
[100] 21742
[101] 21756
test_15
test_3
/usr/bin/docker-current: Error response from daemon: linux runtime spec devices: lstat /dev/dm-25: no such file or directory.
/usr/bin/docker-current: Error response from daemon: linux runtime spec devices: lstat /dev/dm-22: no such file or directory.
test_0
test_33
test_17
test_5
test_9
test_36
test_69
test_6
test_44
test_60
test_13
test_71
test_37
test_79
test_22
test_27
test_72
test_31
test_12
test_38
/usr/bin/docker-current: Error response from daemon: linux runtime spec devices: lstat /dev/dm-24: no such file or directory.
test_96
test_68
test_75
test_7
test_41
test_48
test_46
test_66
test_73
test_40
test_30
test_52
test_55
test_88
test_39
Error response from daemon: Unable to remove filesystem for 6893b25c24356e3d75db938b298f8e07e0847a931c66405dadf6e03804750133: remove /var/lib/docker/containers/6893b25c24356e3d75db938b298f8e07e0847a931c66405dadf6e03804750133/shm: device or resource busy
test_23
test_65
test_62
test_2
/usr/bin/docker-current: Error response from daemon: linux runtime spec devices: lstat /dev/dm-22: no such file or directory.
test_54
test_4
test_26
test_58
test_47
test_77
test_63
test_16
test_24
test_19
test_29
test_74
test_70
test_34
test_57
test_45
test_28
test_11
test_8
test_10
test_49
test_95
test_20
test_32
test_81
test_64
Error response from daemon: Unable to remove filesystem for 35a88372d83353293dcb6b31082a1742e60360898aecbe611473ed8544558d64: remove /var/lib/docker/containers/35a88372d83353293dcb6b31082a1742e60360898aecbe611473ed8544558d64/shm: device or resource busy
test_90
test_25
test_93
test_86
test_80
test_85
test_43
test_67
test_76
test_87
test_14
test_18
test_92
test_50
test_91
test_59
test_89
test_94
test_82
test_78
test_56
/usr/bin/docker-current: Error response from daemon: linux runtime spec devices: lstat /dev/dm-24: no such file or directory.
test_61
test_35
test_53
test_83
test_98
test_42
[1]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[2]   Exit 127                docker run --rm --privileged busybox echo "test_${i}"
[4]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[16]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[40]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[52]   Exit 127                docker run --rm --privileged busybox echo "test_${i}"
[65]   Done                    docker run --rm --privileged busybox echo "test_${i}"
test_97
Error response from daemon: Unable to remove filesystem for 3fff438f7b798f7350ccc673dd38af4dd5cc827debf6696bd297d9f47a89f61c: remove /var/lib/docker/containers/3fff438f7b798f7350ccc673dd38af4dd5cc827debf6696bd297d9f47a89f61c/shm: device or resource busy
test_99
[3]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[6]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[7]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[8]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[10]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[13]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[14]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[18]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[22]   Exit 127                docker run --rm --privileged busybox echo "test_${i}"
[23]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[24]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[28]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[31]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[32]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[34]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[37]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[38]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[39]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[41]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[42]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[45]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[47]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[49]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[53]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[56]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[61]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[63]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[66]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[67]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[69]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[70]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[72]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[73]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[74]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[76]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[80]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[85]   Exit 127                docker run --rm --privileged busybox echo "test_${i}"
[89]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[97]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[98]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[5]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[27]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[55]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[9]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[12]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[17]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[20]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[25]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[29]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[30]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[35]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[46]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[48]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[58]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[59]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[64]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[71]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[75]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[78]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[11]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[15]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[21]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[26]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[33]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[44]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[50]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[68]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[77]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[81]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[82]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[86]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[87]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[88]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[91]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[94]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[96]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[19]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[36]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[51]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[57]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[60]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[62]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[79]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[83]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[90]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[92]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[93]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[95]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[101]+  Exit 127                docker run --rm --privileged busybox echo "test_${i}"
[43]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[54]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[84]   Done                    docker run --rm --privileged busybox echo "test_${i}"
[99]-  Done                    docker run --rm --privileged busybox echo "test_${i}"
[100]+  Done                    docker run --rm --privileged busybox echo "test_${i}"
[root@[REDACTED] andrew.gunnerson]#
```

**Comment by thaJeztah** [comment]
ping @mlaventure any change to get this cherry-picked into docker/runc ?

**Comment by mlaventure** [comment]
@crosbymichael is working on updating `containerd` to use `runc` master, so it'll come with that. It should make it for 17.06

**Comment by thaJeztah** [comment]
pull request for moving runc to master / 1.0-rc is here; https://github.com/moby/moby/pull/33007

**Comment by mlaventure** [comment]
Closing, it was fixed by #33007
